Add op_defaults and rsc_defaults expressions

clumens opened the pull request:
This is an extremely early version of this patch set.  I've done very minimal testing, including not really any on a live cluster.  I just wanted to start getting comments to see if I am on the right track.

With the following XML blurb:
```    <resources>
      <primitive class="stonith" id="Fencing" type="fence_xvm">
        <instance_attributes id="Fencing-instance_attributes">
          <nvpair id="Fencing-instance_attributes-ip_family" name="ip_family" value="ipv4"/>
        </instance_attributes>
        <operations>
          <op id="Fencing-monitor-interval-60s" interval="60s" name="monitor"/>
        </operations>
      </primitive>
      <primitive class="ocf" id="dummy" provider="pacemaker" type="Dummy">
        <instance_attributes id="dummy-instance_attributes">
          <nvpair id="dummy-instance_attributes-op_sleep" name="op_sleep" value="6"/>
        </instance_attributes>
        <operations>
          <op id="dummy-migrate_from-interval-0s" interval="0s" name="migrate_from" timeout="20s"/>
          <op id="dummy-migrate_to-interval-0s" interval="0s" name="migrate_to" timeout="20s"/>
          <op id="dummy-monitor-interval-60s" interval="60s" name="monitor" on-fail="stop"/>
          <op id="dummy-reload-interval-0s" interval="0s" name="reload" timeout="20s"/>
          <op id="dummy-start-interval-0s" interval="0s" name="start" timeout="20s"/>
          <op id="dummy-stop-interval-0s" interval="0s" name="stop" timeout="20s"/>
        </operations>
      </primitive>
    </resources>
    <constraints/>
    <tags/>
    <op_defaults>
      <meta_attributes id="op_defaults-options">
        <nvpair id="op_defaults-options-timeout" name="timeout" value="5s"/>
      </meta_attributes>
      <meta_attributes id="dummy-options">
        <rule id="dummy-options-timeout-rule" score="INFINITY">
          <rsc_expression id="dummy-rule-Dummy" class="ocf" provider="pacemaker" type="Dummy"/>
        </rule>
        <nvpair id="dummy-options-timeout" name="timeout" value="7s"/>
      </meta_attributes>
    </op_defaults>
```

And running that through crm_simulate with enough V args, I am seeing things like this:

```
(parse_op_key)  trace:   Action: monitor
(parse_op_key)  trace:   Resource: Fencing
(make_pairs)    trace: Checking for attributes
(pe_eval_expr)  trace: Testing rule dummy-options-timeout-rule
(pe_eval_rsc_defaults_expr)     trace: Testing rsc_defaults expression: dummy-rule-Dummy
(pe_eval_rsc_defaults_expr)     trace: Class doesn't match: ocf != stonith
(pe_eval_subexpr)       trace: Expression dummy-rule-Dummy failed on all nodes
(pe_eval_expr)  trace: Expression dummy-options-timeout-rule/dummy-rule-Dummy failed
(unpack_attr_set)       trace: Adding attributes from op_defaults-options
(populate_hash)         trace: Setting attribute: timeout = 5s
```

Showing that the rule does not apply for a fencing device, and:

```
(parse_op_key)  trace:   Action: monitor
(parse_op_key)  trace:   Resource: dummy
(make_pairs)    trace: Checking for attributes
(pe_eval_expr)  trace: Testing rule dummy-options-timeout-rule
(pe_eval_rsc_defaults_expr)     trace: Testing rsc_defaults expression: dummy-rule-Dummy
(pe_eval_subexpr)       trace: Expression dummy-rule-Dummy passed on all nodes
(pe_eval_expr)  trace: Rule dummy-options-timeout-rule passed
(unpack_attr_set)       trace: Adding attributes from dummy-options
(populate_hash)         trace: Setting attribute: timeout = 7s
(unpack_attr_set)       trace: Adding attributes from op_defaults-options
```

Showing that it does apply correctly for the dummy resource.

kgaillot commented on xml/options-3.4.rng:
```diff
@@ -0,0 +1,111 @@
+   see upgrade-2.10.xsl
+   - cibtr:table for="cluster-properties"
+   -->
+  <define name="cluster_property_set.nvpair.name-value-unsupported">
```
@jnpkrn , I forget whether we need to keep these "unsupported" blocks once the relevant file moves beyond 3.0 -- were they only needed for the 3.0 schema specifically, or do we need to keep them for all 3.x?

clumens commented on include/crm/pengine/rules_internal.h:
```diff
@@ -24,7 +24,9 @@ crm_time_t *pe_parse_xml_duration(crm_time_t * start, xmlNode * duration_spec);
                       crm_time_t *next_change);
+gboolean pe_eval_op_defaults_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data);
 gboolean pe_eval_role_expr(xmlNode *expr, pe_rule_eval_data_t *rule_data);
+gboolean pe_eval_rsc_defaults_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data);
```
Done.

clumens commented on include/crm/pengine/rules_internal.h:
```diff
@@ -21,6 +21,11 @@ void pe_free_alert_list(GListPtr alert_list);
+int pe_eval_date_expr(xmlNode *expr, pe_rule_eval_data_t *rule_data,
+                      crm_time_t *next_change);
+gboolean pe_eval_role_expr(xmlNode *expr, pe_rule_eval_data_t *rule_data);
+
```
Done.

clumens commented on include/crm/pengine/rules.h:
```diff
@@ -79,12 +79,21 @@ gboolean pe_test_expression(xmlNode *expr, GHashTable *node_hash,
+                                pe_rule_eval_data_t *rule_data, GHashTable *hash,
+                                const char *always_first, gboolean overwrite,
+                                crm_time_t *next_change);
+
```
Done.

clumens commented on include/crm/pengine/internal.h:
```diff
@@ -13,6 +13,7 @@
 #  include <crm/pengine/status.h>
 #  include <crm/pengine/remote_internal.h>
 #  include <crm/common/output.h>
+#  include <crm/pengine/rules.h>
```
This doesn't really make sense to me either.  I guess I can move this new stuff to common.h and just ignore the rest and hope it doesn't break.  I'd really like to understand it, though.

clumens replied:
This also required moving the pe_match_data_t stuff into common.h but I don't think that's a problem.  There's nothing in that structure that's specific to a single file.

kgaillot replied:
In any case we don't need to include rules.h here anymore (status.h includes common.h)

clumens replied:
Done.

kgaillot commented on lib/pengine/utils.c:
```diff
@@ -1009,24 +1051,11 @@ unpack_operation(pe_action_t * action, xmlNode * xml_obj, pe_resource_t * contai
         g_hash_table_replace(action->meta, strdup(field), value_ms);
 
-    } else if (value) {
+    } else if (g_hash_table_lookup(action->meta, field) != NULL) {
```
g_hash_table_remove() will do nothing if nothing matches, so we don't really need a test here anymore

clumens replied:
Done.

clumens replied:
This line somehow ends up being important.  That's what was causing the scheduler tests to fail.

kgaillot replied:
What the heck? That scares me. I don't see how that could possibly matter.

I don't think pacemaker ever inserts hash table entries with a key and a NULL value (the only case where I think the behavior would be different), and even if it did, I would expect a nonexistent key and a NULL value for the key to be treated equivalently by the rest of the code anyway, because we only ever use g_hash_table_lookup() which returns NULL either way.

clumens replied:
We definitely do have keys where the value is NULL (or, we do now with my patches - I guess I'd need to test this against master too):

```
(pe_eval_expr)  trace: Expression Dummy-options-timeout-rule/rule-Dummy failed
(unpack_attr_set)       trace: Adding attributes from op_defaults-options
(populate_hash)         trace: Setting attribute: timeout = 5s
####### interval = (null)
```

The docs for g_hash_table_lookup have a warning for if the value is NULL.  See https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-lookup.

clumens replied:
Verified on master - an fprintf in the same-ish spot also returns keys with a value of NULL.

kgaillot replied:
If you're printing g_hash_table_lookup(), that's most likely a nonexistent key rather than an key that's present with a NULL value. Either way I don't think the rest of the code would treat those cases differently.

I'm at a loss. I keep staring it but I don't see how it can matter.

clumens commented:
> Something went wrong with the monitor interval parsing, see the scheduler tests. Not sure where the issue is.

Looks like it's in e993890.  That patch made sense to me at the time, but not right now.

kgaillot commented:
> Looks like it's in [e993890](https://github.com/ClusterLabs/pacemaker/commit/e99389040e2d035de626f7b865d0c62e43a01519). That patch made sense to me at the time, but not right now.

It makes sense to me too. unpack_operation() is only called by custom_action(), and custom_action() always has the key. I can't think of any situation where the action key has an interval different from the action meta-attributes, but that's the only thing I can see that would be different. I'd be curious if you traced the interval parsed from the key vs the one parsed from the XML to see when it's different.

clumens commented:
I'd be happy to hear other ideas about how I can test this further.

kgaillot commented on lib/pengine/utils.c:
```diff
@@ -1009,24 +1051,11 @@ unpack_operation(pe_action_t * action, xmlNode * xml_obj, pe_resource_t * contai
 
     g_hash_table_remove(action->meta, "id");
 
-    // Normalize interval to milliseconds
```
Let's keep this comment, it's still true -- we're replacing the user-configured interval (in a freeform text format like "10s" or "P10S") with the interval as a number of milliseconds (or removing the interval if it evaluates to 0)

clumens replied:
Done.

kgaillot commented:
We do need to add a scheduler regression test. Come up with a CIB XML that contains several test cases, e.g. something using rsc_defaults, something using op_defaults, something using both, and something using neither. Then we have to get the results to show up in the output graph which means we'll need to trigger some action. Generally I find the easiest way is to set target-role=Stopped on an active resource, then the scheduler will want to stop it.

To add a scheduler test:
1. Put the CIB XML at cts/scheduler/${TEST_NAME}.xml
2. Run "cts/cts-scheduler --update --run $TEST_NAME" and verify the created files are as expected
3. Edit cts/cts-scheduler.in to add test name and description to tests

clumens commented:
This is starting to look pretty good.

> Then we have to get the results to show up in the output graph which means we'll need to trigger some action. Generally I find the easiest way is to set target-role=Stopped on an active resource, then the scheduler will want to stop it.

Just to be clear, I should mark every resource with target-role=Stopped?

kgaillot commented:
> This is starting to look pretty good.
> 
> > Then we have to get the results to show up in the output graph which means we'll need to trigger some action. Generally I find the easiest way is to set target-role=Stopped on an active resource, then the scheduler will want to stop it.
> 
> Just to be clear, I should mark every resource with target-role=Stopped?

More specifically whatever resources you want to trigger a stop action for. The operation timeout etc. values won't show up anywhere in the output unless there's an action scheduled that needs them. With an action scheduled, you can verify in the .exp file that the values are what you want.

kgaillot commented on lib/pengine/rules.c:
```diff
@@ -956,38 +656,44 @@ make_pairs(xmlNode *top, xmlNode *xml_obj, const char *set_name,
  * \param[out] hash          Where to store extracted name/value pairs
  * \param[in]  always_first  If not NULL, process block with this ID first
  * \param[in]  overwrite     Whether to replace existing values with same name
- * \param[in]  now           Time to use when evaluating rules
```
need to document rule_data

clumens replied:
Done.

clumens commented on cts/scheduler/op-rsc-defaults.xml:
```diff
@@ -0,0 +1,103 @@
+          <op id="uses-op_defaults-dummy-monitor-interval-60s" interval="60s" name="monitor" on-fail="stop"/>
+          <op id="uses-op_defaults-dummy-reload-interval-0s" interval="0s" name="reload" timeout="20s"/>
+          <op id="uses-op_defaults-dummy-start-interval-0s" interval="0s" name="start" timeout="20s"/>
+          <op id="uses-op_defaults-dummy-stop-interval-0s" interval="0s" name="stop" timeout="20s"/>
```
I removed all except the monitor operation, and the op_default is now reflected in the .exp file:
```
--- a/cts/scheduler/op-rsc-defaults.exp
+++ b/cts/scheduler/op-rsc-defaults.exp
@@ -63,7 +63,7 @@
     <action_set>
       <rsc_op id="11" operation="start" operation_key="uses-op_defaults-dummy_start_0" on_node="cluster02" on_node_uuid="2">
         <primitive id="uses-op_defaults-dummy" class="ocf" provider="pacemaker" type="Dummy"/>
-        <attributes CRM_meta_name="start" CRM_meta_on_node="cluster02" CRM_meta_on_node_uuid="2" CRM_meta_timeout="20000"  op_sleep="6"/>
+        <attributes CRM_meta_on_node="cluster02" CRM_meta_on_node_uuid="2" CRM_meta_timeout="7000"  op_sleep="6"/>
       </rsc_op>
     </action_set>
     <inputs>
```

clumens commented on cts/scheduler/op-rsc-defaults.exp:
```diff
@@ -0,0 +1,132 @@
+    <action_set>
+      <rsc_op id="10" operation="monitor" operation_key="uses-rsc_defaults-fencing_monitor_60000" on_node="cluster01" on_node_uuid="1">
+        <primitive id="uses-rsc_defaults-fencing" class="stonith" type="fence_xvm"/>
+        <attributes CRM_meta_interval="60000" CRM_meta_name="monitor" CRM_meta_on_node="cluster01" CRM_meta_on_node_uuid="1" CRM_meta_timeout="10000" />
```
Ah yep, I see that.  Is it reasonable to keep it this way in the test?  Perhaps I should change op_defaults-monitor to something else so it doesn't get applied all over the place.

kgaillot replied:
Sure, we want to test a bunch of scenarios. I.e. we should have some action in the graph using the default-default timeout, another using a general op_defaults timeout, another using an op_expression, etc.

Testing rsc_defaults is trickier if none of them appears in the graph. The way to do that would be to set something like target-role or is-managed in the XML so the simulation shows something that needs to be done (e.g. setting target-role=Stopped for a started resource will show a start action, setting is-managed=false for a resource will show "unmanaged" in the status output, etc.)

clumens commented on cts/scheduler/op-rsc-defaults.xml:
```diff
@@ -0,0 +1,103 @@
+        <rule id="rsc_defaults-host_list-rule" score="INFINITY">
+          <rsc_expression id="rsc_defaults-host_list-rule-ping" class="ocf" provider="pacemaker" type="ping"/>
+        </rule>
+        <nvpair id="rsc_defaults-expr-host_list" name="host_list" value="4.2.2.2 4.4.4.4"/>
```
I've replaced this with migration-threshold, and I can see in the log that it's being set:
```
(pe_eval_expr)  trace: Rule rsc_defaults-migration-rule passed
(unpack_attr_set)       trace: Adding attributes from rsc_defaults-migration
(populate_hash)         trace: Setting attribute: migration-threshold = 2
```
but there's nowhere in the cts/schedule/op-rsc-defaults* files that it is reflected, so I'm not sure how much good the test actually does.

clumens replied:
Perhaps I need to use crm_simulate to model failures on the node it starts on to see if it gets moved.

clumens replied:
Okay, one more time.  This feels better to me:
```
$ PCMK_schema_directory=xml/ tools/crm_simulate --xml-file cts/scheduler/op-rsc-defaults.xml --op-inject uses-both-ping_monitor_0@cluster02=1 -O output.xml
$ mv output.xml cts/scheduler/op-rsc-defaults.xml
$ cts/cts-scheduler --update --run op-rsc-defaults
```

kgaillot replied:
I thought all meta-attributes were passed to the graph, but it appears it's just selected ones. Most meta-attributes are only needed by the scheduler itself, the graph is what the controller needs to know. timeout may be the only useful thing to test

clumens commented on cts/scheduler/op-rsc-defaults.exp:
```diff
@@ -0,0 +1,132 @@
+    <action_set>
+      <rsc_op id="9" operation="start" operation_key="uses-rsc_defaults-fencing_start_0" on_node="cluster01" on_node_uuid="1">
+        <primitive id="uses-rsc_defaults-fencing" class="stonith" type="fence_xvm"/>
+        <attributes CRM_meta_on_node="cluster01" CRM_meta_on_node_uuid="1" CRM_meta_timeout="5000" />
```
The family does appear to be set, according to the log:
```
(unpack_resources)      trace: Beginning unpack... <primitive id=uses-rsc_defaults-fencing... >
...
(populate_hash)         trace: Setting attribute: ip_family = ipv6
```
I don't know why it wouldn't be shown on that line, though.  Maybe it gets overridden later (no sign of that in the log)?

Is there any problem with the 5s timeout appearing there, aside from it being an op_default and this resource being named "uses-rsc_defaults"?

kgaillot replied:
ip_family just isn't a meta-attribute

No problem with 5s, it's a good test of op_defaults

clumens commented:
To make this PR maximally confusing, let's start over on the test cases.  I think I was trying to do too much in one file which was just making things more difficult.  I'm pushing a new test case that only handles op_defaults.  Everything here appears to be working except for op-monitor-interval-defaults.  I'm not sure whether that's due to an incorrect test, incorrect testing, or a bug in the code.

kgaillot commented on cts/scheduler/op-defaults.exp:
```diff
@@ -0,0 +1,74 @@
+      <rsc_op id="5" operation="monitor" operation_key="ip-rsc_monitor_0" on_node="cluster02" on_node_uuid="2">
+        <primitive id="ip-rsc" class="ocf" provider="heartbeat" type="IPaddr2"/>
+        <attributes CRM_meta_on_node="cluster02" CRM_meta_on_node_uuid="2" CRM_meta_op_target_rc="7" CRM_meta_timeout="7000" cidr_netmask="32"  ip="172.17.1.1"/>
+      </rsc_op>
```
All of the scheduled actions are 0-second monitors, so they all match op-ping-defaults and get a 7s timeout

None of the resources are being started b/c there's no fence device. If you add one, you should get starts and recurring monitors scheduled, which will let you test the other matches

clumens replied:
I added a fence device and verified all the timeouts match what you've described here, which was also what I intended them to be.  I additionally renamed the ping-related op_default name to op-monitor-defaults, since that's what it does.

clumens commented on cts/scheduler/op-defaults.xml:
```diff
@@ -0,0 +1,81 @@
+          <op_expression id="op-ping-default-expr" name="monitor"/>
+        </rule>
+        <nvpair id="op-ping-monitor-timeout" name="timeout" value="7s"/>
+      </meta_attributes>
```
Done.

clumens commented on cts/scheduler/op-rsc-defaults.xml:
```diff
@@ -0,0 +1,103 @@
+        <rule id="op_defaults-monitor-rule" score="INFINITY">
+          <op_expression id="op_defaults-monitor-expr" name="monitor"/>
+        </rule>
+        <nvpair id="op_defaults-monitor-timeout" name="timeout" value="10s"/>
```
Done.

clumens commented:
Updated with additional tests and documentation.  The one thing I have not tested is what happens when two rules apply to the same resource.  I believe the last one will take effect (and I've updated the documentation to that effect), but I need to add a test for it.  That's coming next.

kgaillot commented on cts/scheduler/rsc-defaults-2.xml:
```diff
@@ -0,0 +1,53 @@
+        <nvpair id="op-unmanaged-nvpair" name="is-managed" value="false"/>
+      </meta_attributes>
+      <meta_attributes id="op-managed">
+        <rule id="op-managed-rule" score="INFINITY">
```
Suddenly I remember the rule score controls the precedence. Try two different scores, it should work automatically. We may already have a regression test for that, not sure.

The case of equal scores is not particularly important, it's on the user to use different scores if they care which one is used.

clumens replied:
Rule score appears to be working, but the higher scoring attribute does not get applied over the lower scoring one:
```
(pe_eval_expr)  trace: Rule op-managed-rule passed
(unpack_attr_set)       trace: Adding attributes from op-managed
(populate_hash)         trace: Setting attribute: is-managed = true
(pe_eval_expr)  trace: Testing rule op-unmanaged-rule
(pe__eval_rsc_expr)     trace: Testing rsc_defaults expression: op-unmanaged-expr
(pe_eval_subexpr)       trace: Expression op-unmanaged-expr passed on all nodes
(pe_eval_expr)  trace: Rule op-unmanaged-rule passed
(unpack_attr_set)       trace: Adding attributes from op-unmanaged
```
is-managed=true is lower scoring than is-managed=false.  Notice that there is no second line where the attribute is actually set.  I guess I've got a bug somewhere.

clumens replied:
I think this is because overwrite is not set... somewhere.  it's kind of hard to track where that happens.  Anyway I guess I would have thought rules just get collected and then the highest scoring gets applied.  At least in this one case, it feels like they all get applied in order of score.

kgaillot commented on doc/Pacemaker_Explained/en-US/Ch-Rules.txt:
```diff
@@ -522,6 +522,121 @@ You may wish to write +end="2005-03-31T23:59:59"+ to avoid confusion.
+|type
+|The name of the resource agent to be matched (required)
+ indexterm:[XML attribute,type attribute,rsc_expression element]
+ indexterm:[XML element,rsc_expression element,type attribute]
```
Are class and type really required? It might be reasonable for someone to set a default timeout for all stonith resources, for example, without specifying a type.

clumens replied:
At the moment they are both required, yes.  You can see this in pe__eval_op_expr.  It would be very easy to change, though.

kgaillot replied:
Let's do that then. I think that will be the final touch

clumens replied:
Okay, just finishing up some testing on this.  The changes will be throughout the patches - XML schema change, docs change, additional test case, changes to the rules code.  I didn't want to make it too easy to review.

clumens replied:
And pushed.

kgaillot commented on doc/Pacemaker_Explained/en-US/Ch-Rules.txt:
```diff
@@ -522,6 +522,121 @@ You may wish to write +end="2005-03-31T23:59:59"+ to avoid confusion.
+ the resource agent; common values include +monitor+, +start+, and +stop+
+ (required).
+ indexterm:[XML attribute,name attribute,op_expression element]
+ indexterm:[XML element,op_expression element,name attribute]
```
OTOH it would be fine to require operation name, it would be more weird to say "all 10-second interval operations of any kind". But it would be just as fine to allow it.

clumens replied:
Yeah, that 's how it is set up now.  Name is required but interval is optional.  If you give interval, both it and name have to match.  Otherwise just name has to match.

kgaillot commented on doc/Pacemaker_Explained/en-US/Ch-Rules.txt:
```diff
@@ -710,6 +825,62 @@ Rules may be used similarly in +instance_attributes+ or +utilization+ blocks.
+        </rule>
+        <nvpair id="op-monitor-timeout" name="timeout" value="7s"/>
+    </meta_attributes>
+</rsc_defaults>
```
op_defaults

clumens replied:
Fixed.

kgaillot commented on doc/Pacemaker_Explained/en-US/Ch-Rules.txt:
```diff
@@ -710,6 +825,62 @@ Rules may be used similarly in +instance_attributes+ or +utilization+ blocks.
+        </rule>
+        <nvpair id="op-monitor-and-timeout" name="timeout" value="8s"/>
+    </meta_attributes>
+</op_defaults>
```
8 seconds :)

clumens replied:
Hah, fixed.  Too much copy and paste going on here.

kgaillot commented on lib/pengine/rules.c:
```diff
@@ -1105,6 +839,481 @@ pe_unpack_versioned_parameters(xmlNode *versioned_params, const char *ra_version
+        return FALSE;
+    }
+
+    if (interval != 0 && interval != rule_data->op_data->interval) {
```
I would use interval_s != NULL instead of interval != 0 ... the configuration could explicitly specify 0 to match only probes

clumens replied:
Done.

kgaillot commented:
Woohoo!
